server/src/main/java/org/elasticsearch/index/search/MatchQuery.java

```diff
@@ -109,20 +109,45 @@ public enum ZeroTermsQuery {
 
     protected final LuceneQueryBuilder.Context context;
     protected final ParsedOptions parsedOptions;
-    protected final Analyzer analyzer;
 
 
     public MatchQuery(LuceneQueryBuilder.Context context, ParsedOptions parsedOptions) {
         this.context = context;
         this.parsedOptions = parsedOptions;
-        this.analyzer = getAnalyzer(parsedOptions.analyzer());
     }
 
-    private Analyzer getAnalyzer(@Nullable String analyzerName) {
-        if (analyzerName == null) {
-            return Lucene.KEYWORD_ANALYZER;
+    /**
+     * Returns the analyzer that was used to index the given column, but throws an exception when the search-analyzer
+     * specified in parse-options is different(from the index-analyzer).
+     */
+    protected Analyzer getSafeAnalyzer(Reference ref) {
+        String searchAnalyzer = parsedOptions.analyzer();
+        if (ref instanceof IndexReference indexRef) {
+            if (searchAnalyzer == null || searchAnalyzer.equals(indexRef.analyzer())) {
+                return context.getAnalyzer(indexRef.analyzer());
+            }
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ENGLISH,
+                    "Column '%s' was indexed with '%s' analyzer, searching with a different analyzer, '%s' is not supported",
+                    indexRef.column(),
+                    indexRef.analyzer(),
+                    searchAnalyzer
+                )
+            );
+        } else {
+            if (searchAnalyzer == null || searchAnalyzer.equals("keyword")) {
+                return Lucene.KEYWORD_ANALYZER;
+            }
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ENGLISH,
+                    "Column '%s' was indexed with keyword analyzer, searching with a different analyzer, '%s' is not supported",
```

Comment:
```suggestion
                    "Column '%s' was indexed with the 'keyword' analyzer, searching with a different analyzer, '%s' is not supported",
```

Add single quotes to match the other error message.